Basic features

gqmelo opened the pull request.

gqmelo committed:
Initial implementation

Mostly inspired by official sentry-plugins repo.

prusse-martin commented on sentry_zendesk/client.py:
```diff
@@ -0,0 +1,60 @@
+        return ticket_id
+
+    def search_tickets(self, query):
+        params = {'query': 'type:ticket subject:{}*'.format(query)}
```
I don't know the proper query language.
Are you sure you want `subject:{}*` and not `subject:*{}*` so you find the "query" in the subject's body middle and not just the start?

gqmelo replied:
Honestly I don't know either :P
Apparently the results does not respect what is stated in the [documentation](https://developer.zendesk.com/rest_api/docs/core/search).

For example the docs says:

> For example, the search results for "top" will include data that match the string "top" as a single word, a single word in a longer phrase ("top tier"), or the prefix of a longer word ("topology")

But neither ```subject:prob``` nor ```subject:"prob"``` return anything.

Some other tests I made:

- ```subject:problema``` returns results like:

```
problema de licencia 
Problemas em Contatos (A/C Rafael Veturrazzi)
suporte solicitado por email - problemas na analise harmônica
Problemas para executar o Fluent 15.0.0 no cluster
Problemas em resultados de simulação com contato
```

- ```subject:problema licencia``` returns results like:

```
Problema de instalacion de licencia
Problema con instalación de licencia
Problemas de reconocimiento de mallas en  fluent
Problemas de licencia
```
- ```subject:problema licen``` returns nothing.
- ```subject:problema licen*``` returns nothing.
- ```subject:problema licenc``` returns some results but much less than ```subject:problema licencia```.
- ```subject:prob``` returns nothing
- ```subject:prob*``` returns results with the word problema anywhere

So, when there is only one word the wildcard makes difference. However with multiple word the wildcard does not help and only querying the whole words seems reliable


```
```

prusse-martin replied:
Since "```subject:problema licen*``` returns nothing." you may want to add it only to one word queries.
Like I will expect ```subject:problema licencia*``` to return nothing as well.

gqmelo replied:
I forgot to show this example. ```subject:problema licencia*``` returns the same as ```subject:problema licencia```. So we can always include the ```*``` for simplicity.

prusse-martin commented on sentry_zendesk/plugin.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        Called by the worker process whenever a new event arrives.
+        """
+        logger.info('Event: {}\nIs new:{}'.format(event, is_new))
```
Just a suggestion. I am not sure I would like ti see 2 lines in the log for this.

gqmelo replied:
Yeah, this is a leftover from debugging sessions. Thanks!

nicoddemus commented on LICENSE:
```diff
@@ -0,0 +1,202 @@
+
+                                 Apache License
```
Why did you chose this license? If not for any special reason I suggest we use `MIT`, which is a straight forward and permissive license.

gqmelo replied:
The [official plugins](https://github.com/getsentry/sentry-plugins) use this license and I took a lot of code snippets from there, so better avoid any copyright problem.
Anyway, the Apache 2.0 license is also very common and is basically like MIT and BSD-3-clause, but with additional protection against patent lawsuits.

nicoddemus replied:
OK! My question was more on the way of "was there a reason for choosing Apache 2?" and that answer is satisfactory enough.

nicoddemus commented on README.rst:
```diff
@@ -0,0 +1,69 @@
+
+Plugin for Sentry which allows linking Zendesk tickets to Sentry issues
+
+    **DISCLAIMER**: Sentry API under development and `is not frozen <https://docs.sentry.io/server/plugins/>`_.
```
This renders [quite ugly](https://github.com/ESSS/sentry-zendesk/tree/basic-features). I suggest removing the indentation.

Also consider moving this under **Limitations**.

gqmelo replied:
I just copied from another plugin, I'll remove the indentation, but I don't think it should be under Limitations. It's a warning, so it should be very visible until the API is frozen.

nicoddemus commented on sentry_zendesk/__init__.py:
```diff
@@ -0,0 +1,13 @@
+
+
+try:
+    VERSION = __import__('pkg_resources') \
```
😢 

Importing `pkg_resources` is quite expensive (like, really). Consider using `setuptools_scm` for automatic versioning instead, see an example [here](https://github.com/ESSS/conda-devenv/pull/39).

gqmelo replied:
Yeah, I agree it is better. I was just following how the official plugins do. I'll change that.

nicoddemus commented on sentry_zendesk/plugin.py:
```diff
@@ -0,0 +1,221 @@
+                group, event, ticket_type='incident', problem_id=problem_id)
+
+    def _get_linked_ticket(self, group):
+        # XXX(dcramer): Sentry doesn't expect GroupMeta referenced here so we
```
Was this copied from somewhere else I guess, do you mean to keep it?

gqmelo replied:
Yeah, this is from the official Jira plugin. I preferred to leave the comment as this seems to be something that they may break in the future.

nicoddemus commented on setup.py:
```diff
@@ -0,0 +1,58 @@
+        'License :: OSI Approved :: Apache Software License',
+        'Operating System :: POSIX :: Linux',
+        "Programming Language :: Python :: 2",
+        'Programming Language :: Python :: 2.7',
```
Do we plan to add Python 3 support soon?

gqmelo replied:
No idea, this depends on [sentry](https://github.com/getsentry/sentry) supporting it.

nicoddemus commented on tox.ini:
```diff
@@ -0,0 +1,42 @@
+
+# Run on a separate env and job so we can run the tests in develop mode, which
+# makes it easier to run coverage and specify the sources.
+[testenv:coverage]
```
Coverage is not on `envlist`, is this intended?

gqmelo replied:
Well, I got that from [pytest](https://github.com/pytest-dev/pytest/blob/master/.travis.yml#L11), but now I realized that pytest excludes it from ```envlist``` because the coverage data upload happens when running ```tox```, so someone could accidentally upload data from a local machine. As here we are uploading the data as a travis step, I'll add it to the ```envlist```.

gqmelo committed:
Automatic deploy to PyPI

In case someone needs to change user/password, here are some docs:

https://docs.travis-ci.com/user/environment-variables/#Encrypting-environment-variables
https://docs.travis-ci.com/user/encryption-keys
https://docs.travis-ci.com/user/deployment/pypi/
